File: src/main/java/calclavia/lib/access/AccessGroup.java
package calclavia.lib.access;

import java.util.LinkedHashSet;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import calclavia.lib.utility.nbt.ISaveObj;

import com.builtbroken.common.Group;

/** Used by a terminal to track what users are part of each group. As well used to setup access
 * points to the terminal.
 * 
 * @author DarkGuardsman */
public class AccessGroup extends Group<AccessUser> implements ISaveObj
{
    protected LinkedHashSet<String> nodes = new LinkedHashSet<String>();
    protected AccessGroup extendGroup;

    public AccessGroup(String name, AccessUser... js)
    {
        super(name, js);
    }

    public void setToExtend(AccessGroup group)
    {
        this.extendGroup = group;
    }

    public AccessUser getMember(String name)
    {
        for (AccessUser user : this.memebers)
        {
            if (user.getName().equalsIgnoreCase(name))
            {
                return user;
            }
        }
        return null;
    }

    @Override
    public void save(NBTTagCompound nbt)
    {
        nbt.setString("groupName", this.getName());
        NBTTagList usersTag = new NBTTagList();
        for (AccessUser user : this.memebers)
        {
            if (!user.isTempary)
            {
                NBTTagCompound accessData = new NBTTagCompound();
                user.save(accessData);
                usersTag.appendTag(accessData);
            }
        }
        nbt.setTag("users", usersTag);
        NBTTagList nodesTag = new NBTTagList();
        for (String str : this.nodes)
        {
            NBTTagCompound accessData = new NBTTagCompound();
            accessData.setString("name", str);
            nodesTag.appendTag(accessData);
        }
        nbt.setTag("nodes", nodesTag);
    }

    @Override
    public void load(NBTTagCompound nbt)
    {
        this.setName(nbt.getString("groupName"));
        NBTTagList userList = nbt.getTagList("users");
        this.getMembers().clear();
        for (int i = 0; i < userList.tagCount(); ++i)
        {
            AccessUser user = AccessUser.loadFromNBT((NBTTagCompound) userList.tagAt(i));
            this.addMemeber(user);
        }
        NBTTagList nodeList = nbt.getTagList("nodes");
        this.nodes.clear();
        for (int i = 0; i < nodeList.tagCount(); ++i)
        {
            this.nodes.add(((NBTTagCompound) nodeList.tagAt(i)).getString("name"));
        }
    }

    public boolean hasNode(String node)
    {
        return this.nodes.contains(node);
    }

    public void addNode(String node)
    {
        this.nodes.add(node);
    }

    public void removeNode(String node)
    {
        if (this.nodes.contains(node))
        {
            this.nodes.remove(node);
        }
    }

    public boolean isMemeber(String string)
    {
        return this.memebers.contains(new AccessUser(string));
    }
}

File: src/main/scala/resonant/lib/access/scala/AccessGroup.scala
package resonant.lib.access.scala

import net.minecraft.nbt.{NBTTagCompound, NBTTagList}
import resonant.lib.access.java.Permission

class AccessGroup extends AbstractAccess
{
  var users = Set.empty[AccessUser]

  def this(nbt: NBTTagCompound)
  {
    this()
    val nbtList = nbt.getTagList("users", 10)
    users = ((0 until nbtList.tagCount()) map (i => new AccessUser(nbtList.getCompoundTagAt(i)))).toSet
    fromNBT(nbt)
  }

  override def toNBT: NBTTagCompound =
  {
    val nbt = super.toNBT

    val userList = new NBTTagList()
    users.foreach(x => userList.appendTag(x.toNBT))
    nbt.setTag("users", userList)
    return nbt
  }

  override def hasPermission(username: String, permission: Permission): Boolean =
  {
    return users.exists(_.username.equals(username)) && (permissions.contains(permission) || users.exists(_.hasPermission(permission)))
  }
}